gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -0,0 +1,841 @@
+  Vector minEigenVector;
+  double minEigenValue = computeMinEigenValue(values, &minEigenVector);
+  return std::make_pair(minEigenValue, minEigenVector);
+}
+
+/* ************************************************************************* */
+template <size_t d>
+bool ShonanAveraging<d>::checkOptimality(const Values &values) const {
+  double minEigenValue = computeMinEigenValue(values);
+  return minEigenValue > parameters_.optimalityThreshold;
+}
+
+/* ************************************************************************* */
+/// Create a tangent direction xi with eigenvector segment v_i
+template <size_t d>
+Vector ShonanAveraging<d>::MakeATangentVector(size_t p, const Vector &v,
+                                              size_t i) {
+  // Create a tangent direction xi with eigenvector segment v_i
+  const size_t dimension = SOn::Dimension(p);
+  const auto v_i = v.segment<d>(d * i);
+  Vector xi = Vector::Zero(dimension);
+  double sign = pow(-1.0, round((p + 1) / 2) + 1);
+  for (size_t j = 0; j < d; j++) {
+    xi(j + p - d - 1) = sign * v_i(d - j - 1);
+    sign = -sign;
+  }
+  return xi;
+}
+
+/* ************************************************************************* */
+template <size_t d>
+Matrix ShonanAveraging<d>::riemannianGradient(size_t p,
+                                              const Values &values) const {
+  Matrix S_dot = StiefelElementMatrix(values);
+  // calculate the gradient of F(Q_dot) at Q_dot
+  Matrix euclideanGradient = 2 * (L_ * (S_dot.transpose())).transpose();
+  // cout << "euclidean gradient rows and cols" << euclideanGradient.rows() <<
+  // "\t" << euclideanGradient.cols() << endl;
+
+  // project the gradient onto the entire euclidean space
+  Matrix symBlockDiagProduct(p, d * nrUnknowns());
+  for (size_t i = 0; i < nrUnknowns(); i++) {
+    // Compute block product
+    const size_t di = d * i;
+    const Matrix P = S_dot.middleCols<d>(di).transpose() *
+                     euclideanGradient.middleCols<d>(di);
+    // Symmetrize this block
+    Matrix S = .5 * (P + P.transpose());
+    // Compute S_dot * S and set corresponding block
+    symBlockDiagProduct.middleCols<d>(di) = S_dot.middleCols<d>(di) * S;
+  }
+  Matrix riemannianGradient = euclideanGradient - symBlockDiagProduct;
+  return riemannianGradient;
+}
+
+/* ************************************************************************* */
+template <size_t d>
+Values ShonanAveraging<d>::LiftwithDescent(size_t p, const Values &values,
+                                           const Vector &minEigenVector) {
+  Values lifted = LiftTo<SOn>(p, values);
+  for (auto it : lifted.filter<SOn>()) {
+    // Create a tangent direction xi with eigenvector segment v_i
+    // Assumes key is 0-based integer
+    const Vector xi = MakeATangentVector(p, minEigenVector, it.key);
+    // Move the old value in the descent direction
+    it.value = it.value.retract(xi);
+  }
+  return lifted;
+}
```

Comment:
@david-m-rosen the retract here is the Cayley map (exponential map is not implemented beyond SO(4))